src/api/MetadataQuery.js

```diff
@@ -0,0 +1,137 @@
+ * @author Box
+ */
+
+import Base from './Base';
+import { CACHE_PREFIX_METADATA, ERROR_CODE_METADATA_QUERY } from '../constants';
+
+class MetadataQuery extends Base {
+    /**
+     * @property {string}
+     */
+    key: string;
+
+    /**
+     * @property {Function}
+     */
+    successCallback: Function;
+
+    /**
+     * @property {Function}
+     */
+    errorCallback: ElementsErrorCallback;
+
+    /**
+     * Creates a key for the metadata cache
+     *
+     * @param {string} id - metadata template
+     * @return {string} key
+     */
+    getCacheKey(id: string): string {
+        return `${CACHE_PREFIX_METADATA}query_${id}`;
+    }
+
+    /**
+     * API URL for metadata query
+     * @return {string} base url for files
+     */
+    getUrl(): string {
+        return `${this.getBaseApiUrl()}/metadata_queries/execute`;
+    }
+
+    /**
+     * @param {Object} response
+     */
+    queryMetadataSuccessHandler = ({ data }: { data: Object }): void => {
+        const cache: APICache = this.getCache();
+        cache.set(this.key, data);
+        this.finish();
+    };
+
+    /**
+     * Does the network request to metadata query API
+     * @param {Object} query query object with SQL Clauses like properties
+     * @return {Promise}
+     */
+    queryMetadataRequest(query: Object): Promise<void> {
+        if (this.isDestroyed()) {
+            return Promise.reject();
+        }
+
+        this.errorCode = ERROR_CODE_METADATA_QUERY;
+        return this.xhr
+            .post({
+                url: this.getUrl(),
+                data: query,
+            })
+            .then(this.queryMetadataSuccessHandler)
+            .catch(this.errorHandler);
+    }
+
+    /**
+     * Returns true for cache hit for metadata query results
+     *
+     * @return {boolean} if query results are loaded
+     */
+    isLoaded(): boolean {
+        const cache: APICache = this.getCache();
+        return cache.has(this.key);
+    }
+
+    /**
+     * Returns the results
+     *
+     * @return {void}
+     */
+    finish(): void {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        const cache: APICache = this.getCache();
+        const metadataQueryData = cache.get(this.key);
+        this.successCallback(metadataQueryData);
+    }
+
+    /**
+     * API for querying enterprise metadata
+     * @param {Object} query - metadata query object
+     * @param {Function} successCallback - Function to call with results
+     * @param {Function} errorCallback - Function to call with errors
+     * @param {boolean|void} [options.forceFetch] - Bypasses the cache
+     * @return {void}
+     */
+    queryMetadata(
+        query: Object,
+        successCallback: Function,
+        errorCallback: ElementsErrorCallback,
+        options: Object = {},
+    ): void {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        const { context = {} } = options;
+        this.key = this.getCacheKey(context.id);
```

Comment:
What is the `context` and `id` represent here? Does it make sense to do something similar to `Search.js` where the key is the encoded query?

Reply:
The idea is `context` object will have some useful info about the query thats hard to extract from query object. e.g.enterprise_id, template_key, ancestor_folder_id etc. This is different from the `Search.js` where the query is a simple `string`. 
Component using this API should do the extraction and pass down this info for context. `id` could be generated as a unique cache property for each query.